File: lab16/LabData/tabDeliLoader.vb
'
' Created by SharpDevelop.
' User: Angel Salamanca
' Date: 24/12/2015
' Time: 13:21
' 
' To change this template use Tools | Options | Coding | Edit Standard Headers.
'

Imports System.IO
Imports System.Collections.Generic 

Public Class tabDeliLoader
	
	Private mySep As Char
	Private DS As TablasLab
	Private fileName As String
	Private varList As List(Of String)
	Private sr As StreamReader 
	Private opResult As loadResult 
	Private typeList As Char(,)
	Private strValues As Dictionary(Of String, List(Of String))
	Private intValues As Dictionary(Of String, List(Of Int32))
	Private contValues As Dictionary(Of String, List(Of Double))
	Private readOnly maxTextItems As Int32 = 100
	Private ReadOnly maxIntItems As Int32 = 100
    Private ReadOnly maxContItems As Int32 = 10
    Private strLength As Int32 = 30
    Private guessedRecordLength As Int32

    Public Enum loadResult
		Success
		DuplicateVar
	End Enum
	
	Public SUb New(byref newDS As TablasLab, sep As Char, byref fn As String)
		Me.mySep  =sep
		Me.DS = newDS
		Me.fileName = fn
		
	End Sub
	
	Public Sub Load
		Me.strValues = New Dictionary(Of String, List(Of String))
		Me.intValues = New Dictionary(Of String, List(Of Int32))
		Me.contValues = New Dictionary(Of String, List(Of Double))
		
		Me.sr =New StreamReader(Me.fileName)
		
		If readHeader Then
			If readLines Then
				populateDS
			End If
		End If
		
		Sr.close
		
		
	End Sub
	
	Private Function readHeader() As boolean
		Dim words() As String = Me.getWords()
		Me.varList = New List(Of String)
		
		For Each varName As String In words
			Dim properVarName As String =  MLBase.XmlTools.To2007Name(varName)
			If varList.Contains(properVarName) Then
				Me.opResult  = loadResult.DuplicateVar 
				return False 
			End If
			varList.Add(properVarName )
			
		Next
		Return True
	End Function
	
	Private Function readLines As Boolean

        Dim myInt As Int32, myDouble As Double
        Me.guessedRecordLength = 0

        Me.typeList = Array.CreateInstance(GetType(Char), 10, varList.Count)
        For nl As Int32 = 1 To 10 ' read 10 lines
            Dim words() As String = Me.getWords(True)
            ' go over words
            For nw As Int32 = 0 To Math.Min(words.Length, varList.Count) - 1
                Dim varName As String = Me.varList(nw)
                ' Is it integer?
                If Int32.TryParse(words(nw), NumberStyles.Any, Unica.Entorno.EngineCUlture, myInt) Then
                    typeList(nl - 1, nw) = "N"
                Else
                    If Double.TryParse(words(nw), NumberStyles.Any, Unica.Entorno.EngineCUlture, myDouble) Then
                        typeList(nl - 1, nw) = "C"
                    Else
                        typeList(nl - 1, nw) = "T"
                    End If
                End If
            Next nw
        Next nl
        Me.guessedRecordLength = Me.guessedRecordLength / 10


        ' Decide of var type
        For nv As Int32 = 0 To Me.varlist.Count-1
			Dim varType As String = "N" ' Default
			For nl As Int32 = 0 To Me.typeList.GetLength(0) - 1
				If typeList(nl, nv) = "T" Then
					varType = "T"
					Exit for
				End If
				If typeList(nl, nv) = "C" Then
					varType = "C"
				End If
			Next
			typeList(0, nv) = varType 
			
		Next
		
		Return true
	End Function
	
	Private Function populateDS As Boolean
		
		Dim initialPos As Int32 = 1
		
		For nv As Int32 = 0 To Me.varList.Count - 1
			Dim myLen As Int32
			Select Case Me.typeList(0, nv)
				Case "N"
					myLen = MLStudioLib.Constants.intDefaultLength 
				Case "C"
					myLen = MLStudioLib.Constants.contDefaultLength 
				Case Else
                    myLen = Me.strLength
            End Select
			Dim oldVar As Boolean = DS.NuevaVar(Me.varList(nv), Me.typeList(0, nv), initialPos,	 myLen, MLBase.MLBase.FiltroView.Entrada, False)
			initialPos += myLen
			
			If Not oldVar Then
				Select Case Me.typeList(0, nv)
				Case "N"
					Me.intValues.Add(Me.varList(nv), New List(Of Int32))
				Case "C"
					Me.contValues.Add(Me.varList(nv), New List(Of Double))
				Case "T"
					Me.strValues.Add(Me.varList(nv), New List(Of String))
				End Select
			End if	
			
		Next
		DS.LongitudRegistro = Math.Max(DS.LongitudRegistro, initialPos - 1)  ' In case we have -Old vars that go beyond
		
		DS.CreaSinGrupo()
        DS.CreaGruposBIM()

        DS.CreaSinCat()
        DS.CatSinGrupo()
		
		Return true
	End Function

    Private Function getWords(Optional ByVal addToLength As Boolean = False) As String()
        Dim L As String = Me.sr.ReadLine
        If addToLength Then Me.guessedRecordLength += L.Length

        Return L.Split(mySep)
    End Function

    Public Sub loadData()
		Dim HayCRLF As Boolean = True
        Dim Leido, Campo, L As String, LC() As Char
        Dim Leidos As Integer = 1, Nregs As Integer = 0
        Dim sb As New StringBuilder, varRow As datarow
        
        Dim Il As New InputLab(DS, DS.ArchivoOriginal, False, True, False)
        Il.AbrirEscritura()

        sr.Close
        Me.sr =New StreamReader(Me.fileName)
        sr.ReadLine ' skip header

        Unica.HojaPrincipal.ShowProgressBar(GM("FNP0017"), 1, sr.BaseStream.Length \ Me.guessedRecordLength)

        Do While sr.Peek > -1
            Unica.HojaPrincipal.PasoFinoAdelante()
            Dim words() As String = Me.getWords
            sb.Clear 
            
             Dim Col As Integer
             For I As Int32 = 0 To  varList.Count-1
             	
             	varRow = DS.BuscarVar(Me.varList(I))
             	Dim varName As String = varRow("Nombre")
             	Dim Longitud As Integer = varRow("Longitud")


                If I < words.Length Then
                    Leido = words(I)
                    If Leido.Length > Me.strLength Then Leido = Leido.Substring(0, Me.strLength)
                Else
                    	Leido = ""
					End If
                    Select Case CStr(varRow("Tipo"))
                        Case "T"
                            
                            Campo = Leido.PadRight(Longitud)
                            If Me.strValues.Keys.Contains(varName)
                            	Dim myList As List(Of String) = Me.strValues.Item(varName)
                            	If Not myList.Contains(Campo) AndAlso myList.Count < 1 + maxTextItems  Then
                            		myList.add(New String(Campo))
                            	End If
                            End if 
                        Case "N", "C"
                        	campo = Leido.PadLeft(Longitud)
                        	
                        	If varRow("Tipo") = "N" Then
                        		Dim intValue As Int32
                        		Int32.TryParse(Campo, NumberStyles.Any, Unica.Entorno.EngineCUlture, intValue)
                        		If Me.intValues.Keys.Contains(varName) then
                        			Dim myList As List(Of Int32) = Me.intValues.Item(varName)
                            		If Not myList.Contains(intValue) AndAlso myList.Count < 1 + maxContItems  Then
                            			myList.add(intValue)
                            		End If
                            	End if	
                        	Else
                        		Dim contValue As Double
                        		Double.TryParse(Campo, NumberStyles.Any, Unica.Entorno.EngineCUlture, contValue)
                        		If Me.contValues.Keys.Contains(varName) then
                        			Dim myList As List(Of Double) = Me.contValues.Item(varRow("Nombre"))
                            		If Not myList.Contains(contValue) AndAlso myList.Count < 1 + Me.maxContItems Then
                            			myList.add(contValue)
                            		End If
                            	End if	
                        	End If
                        Case "F"
                        	Throw New Exception 
                    End Select
                    Sb.Append(Campo)
             Next
            If sb.Length < DS.LongitudRegistro Then sb.Append(New String(" ", DS.LongitudRegistro - sb.Length))
            Il.Escribir(System.Text.Encoding.Default.GetBytes(sb.ToString))
        Loop

        sr.Close()

        Unica.RL.RegTrazas("loadData", "All records loaded")

        Unica.HojaPrincipal.HideProgressBar()

		Il.CerrarEscritura()

		Me.autoCatGeneration 

        Unica.HojaPrincipal.ProcessCompleted()
	End Sub
	
	Private Sub autoCatGeneration()
		
		Dim varRow As DataRow 
		
		' String first
		For Each kp As keyvaluepair(Of String , List(Of String)) In Me.strValues 
			varRow = DS.BuscarVar(kp.Key)
			If kp.Value.Count <= Me.maxTextItems Then
				kp.Value.Sort 
				For Each catName As String In kp.Value 
					Dim catRow As DataRow = DS.CreaCat(varRow, catName)
				Next
				
			End If
			
		Next
		
		' Int now 
		For Each kp As keyvaluepair(Of String , List(Of Int32)) In Me.intValues  
			varRow = DS.BuscarVar(kp.Key)
			If kp.Value.Count <= Me.maxIntItems  Then
				kp.Value.Sort 
				For Each catValue As Int32 In kp.Value 
					Dim catRow As DataRow = DS.CreaCat(varRow, catValue.ToString(Unica.Entorno.EngineCUlture))
					catRow("Desde") = catValue
					catRow("Hasta") = catValue
				Next
				
			End If
			
		Next
		
		' And Cont
		For Each kp As keyvaluepair(Of String , List(Of Double)) In Me.contValues  
			varRow = DS.BuscarVar(kp.Key)
			If kp.Value.Count <= Me.maxContItems Then
				kp.Value.Sort 
				For Each catvalue As Double In kp.Value 
					Dim catRow As DataRow = DS.CreaCat(varRow, catValue.ToString(Unica.Entorno.EngineCUlture))
					catRow("Desde") = catValue
					catRow("Hasta") = catValue
				Next
				
			End If
			
		Next
	End Sub
	
End Class

File: lab16/LabData/tabDeliLoader.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using LabQuant;
using System.IO;
using System.Globalization;

namespace LabData
{
    public class tabDeliLoader
    {
        private analyticalDictionary myDict;
        private Char mySep;
        private String fileName, outFileName;
        private Dictionary<String, List<String>> strValues;
        private Dictionary<String, List<Int32>> intValues;
        private Dictionary<String, List<Double>> contValues;
        private StreamReader sr;
        private List<String> varList;
        private Int32 guessedRecordLength;
        private loadResult opResult;
        private General.typeOfVariable[,] typeList;
        private CultureInfo myCulture;
        private List<varColumn> columns;
        private readonly int maxTextItems = 100;
        
        public enum loadResult { Success = 0, DuplicateVar }

        public tabDeliLoader(string dName, string dLabPath, string fName, string outfName, Char sep, CultureInfo dCulture)
        {
            myDict = new analyticalDictionary(dName, dLabPath);
            mySep = sep;
            fileName = fName;
            outFileName = outfName;
            myCulture = dCulture;
        }

        public analyticalDictionary  Load()
        {
            strValues = new Dictionary<String, List<String>>();
            intValues = new Dictionary<String, List<Int32>>();
            contValues = new Dictionary<String, List<Double>>();

            sr = new StreamReader(fileName);

            if (readHeader())
            {
                if (readLines())
                {
                    populateDS();
                }
            }

            sr.Close();
            return this.myDict;
        }

        // build a list of variable names. Must be unique.
        private bool readHeader()
        {
            List<String> words = this.getWords();
            varList = new List<String>();


            foreach (String varName in words)
            {
                var properVarName = General.toLabName(varName);
                if (varList.Contains(properVarName))
                {
                    this.opResult = loadResult.DuplicateVar;
                    return false;
                }
                varList.Add(properVarName);
            }
            return true;
        }

        // Guess variable type scanning first 10 records
        private Boolean readLines()
        {

            int myInt;
            double myDouble;
           

            this.guessedRecordLength = 0;
            typeList = (General.typeOfVariable[,])Array.CreateInstance(typeof(General.typeOfVariable), 10, varList.Count);

            for (int nl = 1; nl <= 10; nl++) // read 10 lines
            {
                List<String> words = this.getWords(true);
                // go over words
                for (int nw = 0; nw < Math.Min(words.Count, varList.Count); nw++)
                {
                    var varName = this.varList[nw];
                    // Is it integer?
                    if (int.TryParse(words[nw], NumberStyles.Any, myCulture, out myInt))
                    {
                        typeList[nl - 1, nw] = General.typeOfVariable.Integer;
                    }
                    else
                    {
                        if (Double.TryParse(words[nw], NumberStyles.Any, myCulture, out myDouble))
                        {
                            typeList[nl - 1, nw] = General.typeOfVariable.Double ;
                        }
                        else
                        {
                            typeList[nl - 1, nw] = General.typeOfVariable.String;
                        }
                    }
                } //nw
        } // nl
        this.guessedRecordLength = this.guessedRecordLength / 10;
            
        // Decide of var type
        for (int nv = 0; nv < varList.Count; nv++)
            {
                General.typeOfVariable varType = General.typeOfVariable.Integer ; // Default
                for (int nl = 0; nl < typeList.GetLength(0) - 1; nl++)
                { 
				if (typeList[nl, nv] == General.typeOfVariable.String )
                    {
                       varType = General.typeOfVariable.String;
                       break;
                    }
               if (typeList[nl, nv] == General.typeOfVariable.Double )
                    {
                        varType = General.typeOfVariable.Double;
                    }
                
            }
                typeList[0, nv] = varType;
            }
        
        return true;
    }
        private List<String> getWords(Boolean addToLength = false)
        {
            var L = sr.ReadLine();
            if (addToLength)
            {
                this.guessedRecordLength += L.Length;
            }
            return L.Split(mySep).ToList();
        }

        private Boolean  populateDS()
        { 
        
		for (int nv = 0; nv < varList.Count; nv++)
            {
                
            var oldVar = myDict.addVar(varList[nv], typeList[0, nv], General.blockType.Input);

                if (oldVar != null)
                {
                    switch (typeList[0, nv])
                    {
                        case General.typeOfVariable.Integer:
                            intValues.Add(varList[nv], new List<Int32>());
                            break;
                    case General.typeOfVariable.Double:
                            contValues.Add(varList[nv], new List<Double>());
                            break;
                    case General.typeOfVariable.String:
                            strValues.Add(varList[nv], new List<String>());
                            break;
                    }
                }

        }
		
		return true;
	}

        public void loadData()
        {
            columns = new List<varColumn>();
            int intValue = Int32.MinValue ;
            double doubleValue = Double.NaN ;
            
            // Create dir if needed
            Directory.CreateDirectory(this.myDict.filePath(this.outFileName));
            // init columns
            for (int i = 0; i < this.varList.Count; i++)
            {
                string varName = this.varList[i];
                var var = this.myDict.varByName[varName];
                //string filename = this.myDict.filePath(this.outFileName + Path.DirectorySeparatorChar) + varName + General.varFileSuffix;
                columns.Add(new varColumn(var, this.myDict.filePath(this.outFileName + Path.DirectorySeparatorChar)));
            }

            sr = new StreamReader(this.fileName);
            sr.ReadLine(); // skip header

            // init progressbar

            while (sr.Peek()>-1)
            {
                // fine progress step
                var words = this.getWords();

                for (int i= 0; i< this.varList.Count; i++)
                {
                    string varName = this.varList[i];
                    var var = this.myDict.varByName[varName];
                    // get the value
                    switch (var.variableType)
                    {
                        case General.typeOfVariable.String:
                            if (i >= words.Count)
                                columns[i].addStringValue(string.Empty);
                            else
                                columns[i].addStringValue(words[i]);
                            break;
                        case General.typeOfVariable.Integer:
                        case General.typeOfVariable.Date:
                            if (i >= words.Count)
                                columns[i].stringValues.Add(String.Empty);
                            else
                                Int32.TryParse(words[i], NumberStyles.Any, this.myCulture, out intValue);
                                columns[i].intValues.Add(intValue);
                            break;
                        case General.typeOfVariable.Double:
                            if (i >= words.Count)
                                columns[i].stringValues.Add(String.Empty);
                            else
                                Double.TryParse(words[i], NumberStyles.Any, this.myCulture, out doubleValue);
                                columns[i].doubleValues.Add(doubleValue);
                            break;
                    }
                    
                }
            }

            sr.Close();

            foreach (varColumn varcol in columns)
                varcol.write();

            this.autoCatGeneration();
     
    }
    private void autoCatGeneration()
        {
   
            // String first
            foreach (varColumn varCol in this.columns)

            {
                switch (varCol.myVariable.variableType)
                {
                    case General.typeOfVariable.String:
                        if (varCol.stringCount.Count <= this.maxTextItems)
                        {
                            var strVals = varCol.stringCount.Keys.ToList();
                            strVals.Sort();
                            foreach (string strVal in strVals)
                            {
                                if (strVal != String.Empty)
                                {
                                    this.myDict.addCat(varCol.myVariable, strVal, strVal); // do not create empty Cat, missing is just as good
                                }
                            }
                         }
                        break;
                    case General.typeOfVariable.Integer:
                    case General.typeOfVariable.Date:
                    case General.typeOfVariable.Double:
                        break;
                }
            }
                
            /*
            ' Int now 
                    For Each kp As keyvaluepair(Of String, List(Of Int32)) In Me.intValues
                        varRow = DS.BuscarVar(kp.Key)
                        If kp.Value.Count <= Me.maxIntItems  Then
                            kp.Value.Sort
                            For Each catValue As Int32 In kp.Value
                                Dim catRow As DataRow = DS.CreaCat(varRow, catValue.ToString(Unica.Entorno.EngineCUlture))
                                catRow("Desde") = catValue
                                catRow("Hasta") = catValue
                            Next

                        End If

                    Next

                    ' And Cont
                    For Each kp As keyvaluepair(Of String, List(Of Double)) In Me.contValues
                        varRow = DS.BuscarVar(kp.Key)
                        If kp.Value.Count <= Me.maxContItems Then
                            kp.Value.Sort
                            For Each catvalue As Double In kp.Value
                                Dim catRow As DataRow = DS.CreaCat(varRow, catValue.ToString(Unica.Entorno.EngineCUlture))
                                catRow("Desde") = catValue
                                catRow("Hasta") = catValue
                            Next

                        End If
                        }
            */
        }




    }
}
